src/providers/krb5/krb5_child.c

```diff
@@ -487,6 +499,63 @@ static krb5_error_code tokeninfo_matches_pwd(TALLOC_CTX *mem_ctx,
     return 0;
 }
 
+static krb5_error_code
+tokeninfo_matches_oauth2(TALLOC_CTX *mem_ctx,
+                         const krb5_responder_otp_tokeninfo *ti,
+                         const char *pwd,
+                         size_t len,
+                         char **out_token,
+                         char **out_pin)
+{
+    TALLOC_CTX *tmp_ctx;
+    char *oauth2_url;
+    char *oauth2_pin;
+    char *token;
+    errno_t ret;
+
+    tmp_ctx = talloc_new(NULL);
+    if (tmp_ctx == NULL) {
+        return ENOMEM;
+    }
+
+    if (ti->flags & KRB5_RESPONDER_OTP_FLAGS_NEXTOTP) {
+        ret = ENOTSUP;
+        goto done;
+    }
+
+    ret = otp_parse_oauth2_challenge(tmp_ctx, ti->challenge, &oauth2_url,
+                                     &oauth2_pin);
+    if (ret == ENOENT) {
+        ret = ENOTSUP;
+        goto done;
+    } else if (ret != EOK) {
+        goto done;
+    }
+
+    token = talloc_strndup(tmp_ctx, pwd, len);
+    if (token == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+    talloc_set_destructor(token, token_pin_destructor);
+
+    if (strlen(oauth2_pin) != len || strncmp(oauth2_pin, token, len) != 0) {
```

Comment:
Sorry for the ignorant question, but where does the `len` value come from here? I am trying to understand the reasoning for comparing/matching the values here.

Reply:
So we compare that whatever pin is stored in the challenge is same as the pin given in input authtoken.